File: 98-validate-binary-search-tree/validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        def helper(root,minv,maxv):
            if root is None:
                return True
            if root.val <=minv or root.val >= maxv:
                return False
            left = helper(root.left,minv,root.val)
            right= helper(root.right,root.val,maxv)
            return left and right

        return helper(root,float('-inf'),float('inf'))
        
        

File: 98-validate-binary-search-tree/validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return helper(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }

    public boolean helper(TreeNode root, long minv, long maxv){
        if(root == null){
            return true;
        }

        if (root.val <= minv || root.val >= maxv){
            return false;
        }

        boolean left = helper(root.left, minv, root.val);
        boolean right = helper(root.right, root.val, maxv);

        return left && right;
    }
}